src/jubjub/mod.rs

```diff
@@ -434,6 +460,260 @@ impl JubjubBls12 {
+/// the file differs from that being requested, we return `true` to indicate that a new exp-table
+/// must be generated and written.
+fn must_write_exp_table(
+    n_segments_required: usize,
+    window_size_required: u32,
+    exp_table_path: &str,
+) -> bool {
+    let mut file = match File::open(exp_table_path) {
+        Ok(file) => file,
+        _ => return true,
+    };
+
+    let n_segments_in_file = {
+        let mut buf = [0u8; 4];
+        if file.read_exact(&mut buf).is_err() {
+            return true;
+        }
+        u32::from_le_bytes(buf) as usize
+    };
+
+    if n_segments_in_file < n_segments_required {
+        return true;
+    }
+
+    let window_size_in_file = {
+        file.seek(SeekFrom::Start(4)).unwrap();
+        let mut buf = [0u8; 4];
+        if file.read_exact(&mut buf).is_err() {
+            return true;
+        }
+        u32::from_le_bytes(buf)
+    };
+
+    window_size_in_file != window_size_required
+}
+
+/// Create the exp-table for the provided set of generators.
+fn create_exp_table_for_generators(
+    generators: &[edwards::Point<Bls12, PrimeOrder>],
+    window_size: u32,
+) -> Vec<Vec<Vec<edwards::Point<Bls12, PrimeOrder>>>> {
+    generators
+        .par_iter()
+        .cloned()
+        .map(|mut generator| {
+            let mut tables = vec![];
+
+            let mut n_bits = 0;
+            while n_bits <= fs::Fs::NUM_BITS {
+                let mut table = Vec::with_capacity(1 << window_size);
+                let mut base = edwards::Point::zero();
+
+                for _ in 0..(1 << window_size) {
+                    table.push(base.clone());
+                    base = base.add(&generator, &JUBJUB_BLS12_PARAMS_WITH_EDWARDS_D);
+                }
+
+                tables.push(table);
+                n_bits += window_size;
+
+                for _ in 0..window_size {
+                    generator = generator.double(&JUBJUB_BLS12_PARAMS_WITH_EDWARDS_D);
+                }
+            }
+
+            tables
+        })
+        .collect()
+}
+
+/// Given a path to an exp-table file, this function creates any directories in that path that do
+/// not exist.
+fn create_exp_table_path_path_dirs_if_dne(path: &str) -> io::Result<()> {
+    let path = Path::new(path);
+    let dirs = path.parent().expect("invalid exp-table path");
+    let path_is_just_file_name = dirs.to_str().unwrap().is_empty();
+
+    if path_is_just_file_name {
+        Ok(())
+    } else {
+        create_dir_all(dirs)
+    }
+}
+
+/// Generates the Pedersen Hash parameter's exp-table in chunks of segments, writing each chunk of
+/// the exp-table to the file at `exp_table_path`. This function generates the exp-table in such a
+/// way that there is never more than `MAX_EXP_TABLE_SEGMENTS_IN_MEMORY` segments worth of the
+/// exp-table in memory at a given time.
+fn write_exp_table_to_file(
+    generators: &[edwards::Point<Bls12, PrimeOrder>],
+    window_size: u32,
+    exp_table_path: &str,
+) {
```

Comment:
it is probably better if we return a `io::Result` here, so that there is some way of handling failed file reads/writes